### PR TITLE
Update metrics for .NET 8 changes

### LINKED ISSUE #36750: [Metrics] Metrics naming standard
.NET built-in metrics follow OTel's naming standard. .NET metrics docs should recommend this format to people writing their own metrics in https://learn.microsoft.com/en-us/dotnet/core/diagnostics/metrics-instrumentation#create-a-custom-metric

Writing down a standard will create internal consistency between .NET's built-in counters. And consistency in the community counters.

The standard, off the top of my head:

* Meter name:
  * The NuGet package name/assembly name. Or a namespace that the metrics belong to if the library/app has many metrics.
  * Use namespace naming standard: Pascal case, dot-seperated.
* Counter name:
  * Use the OTel semantic conventions naming standard. Lowercase, dot-separated sections for categories, underscores for spaces.
* Counter unit:
  * UCUM format. Non-histogram counters should place the unit in curly braces, e.g. `{requests}`.
* Counter tags:
  * Use the OTel semantic conventions naming standard. Lowercase, dot-separated sections for categories, underscores for spaces.

I think the OTel semantic conventions have some content we can follow.

### PR DESCRIPTION
**Pull request by noahfalk**
- Adds instructions for using IMeterFactory with Dependency Injection (#36749)
- Adds testing examples for MetricCollector (#36749)
- Updates naming guidance to use OpenTelemetry conventions (Fixes #36750)

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/diagnostics/metrics-instrumentation.md](https://github.com/dotnet/docs/blob/ac14f40d6c392ba7fca72ab380b999965e0f95bd/docs/core/diagnostics/metrics-instrumentation.md) | [Creating Metrics](https://review.learn.microsoft.com/en-us/dotnet/core/diagnostics/metrics-instrumentation?branch=pr-en-us-36781) |

### CODE REVIEW

**Comment by noahfalk** [comment]
@IEvangelist - I'm seeing lint errors but if I understand them correctly they are in other files that this PR isn't touching. Am I supposed to fix those too?

**Comment by noahfalk** [comment]
cc @tarekgh @JamesNK 
James, looking at the text I think its probably good that ASP.NET docs have more ASP.NET-specific examples but happy to get your feedback on any of it.

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (gewarren) [inline docs/core/diagnostics/metrics-instrumentation.md:81]:
```suggestion
[OpenTelemetry naming guidelines](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#general-guidelines),
```

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (gewarren) [inline docs/core/diagnostics/metrics-instrumentation.md:82]:
```suggestion
which use dotted hierarchical names. Assembly names or namespace names for code being instrumented are usually a good choice. If an assembly adds instrumentation
```

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (gewarren) [inline docs/core/diagnostics/metrics-instrumentation.md:86]:
```suggestion
[OpenTelemetry naming guidelines](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#general-guidelines), which use dotted hierarchical names
```

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (gewarren) [inline docs/core/diagnostics/metrics-instrumentation.md:87]:
```suggestion
and an underscore ('_') as the separator between multiple words in the same element. Not all metric tools preserve the Meter name as part of the final metric name, so it's beneficial
```

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (gewarren) [inline docs/core/diagnostics/metrics-instrumentation.md:122]:
```suggestion
## Get a Meter via dependency injection
```

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (gewarren) [inline docs/core/diagnostics/metrics-instrumentation.md:124]:
```suggestion
In the previous example, the Meter was obtained by constructing it with `new` and assigning it to a static field. Using statics this way is not a good approach when using dependency
```

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (gewarren) [inline docs/core/diagnostics/metrics-instrumentation.md:125]:
```suggestion
injection (DI). In code that uses DI, such as ASP.NET Core or apps with [Generic Host](../extensions/generic-host.md), create the Meter object using
```

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (gewarren) [inline docs/core/diagnostics/metrics-instrumentation.md:126]:
```suggestion
<xref:System.Diagnostics.Metrics.IMeterFactory>. Starting in .NET 8, hosts will automatically register <xref:System.Diagnostics.Metrics.IMeterFactory> in the service container
```

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (gewarren) [inline docs/core/diagnostics/metrics-instrumentation.md:128]:
```suggestion
The meter factory integrates metrics with DI, keeping Meters in different service collections isolated from each other even if they use an identical name. This is
```

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (gewarren) [inline docs/core/diagnostics/metrics-instrumentation.md:129]:
```suggestion
especially useful for testing so that multiple tests running in parallel only observe measurements produced from within the same test case.
```

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (gewarren) [inline docs/core/diagnostics/metrics-instrumentation.md:175]:
```suggestion
Inject the metrics type and record values where needed. Because the metrics type is registered in DI, it can be used with MVC controllers, minimal APIs, or any other type that is created by DI:
```

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (gewarren) [inline docs/core/diagnostics/metrics-instrumentation.md:528]:
```suggestion
## Test custom metrics
```

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (gewarren) [inline docs/core/diagnostics/metrics-instrumentation.md:533]:
```suggestion
### Test with dependency injection
```

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (gewarren) [inline docs/core/diagnostics/metrics-instrumentation.md:535]:
```suggestion
The following code shows an example test case for code components that use dependency injection and IMeterFactory.
```

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (gewarren) [inline docs/core/diagnostics/metrics-instrumentation.md:572]:
```suggestion
### Test without dependency injection
```

**Review by gewarren — APPROVED** [review]

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (tarekgh) [inline docs/core/diagnostics/metrics-instrumentation.md:562]: >collection [](http://example.com/codeflow?start=26&length=10)

nit: will be better calling it `servceCollection`.

**Review by tarekgh — APPROVED** [review]
LGTM

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (gewarren) [reply docs/core/diagnostics/metrics-instrumentation.md:562]: But don't forget the `i`.

**Comment by gewarren** [comment]
I will try to fix the lint errors separately.

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (tarekgh) [reply docs/core/diagnostics/metrics-instrumentation.md:562]: I fixed it :-)

**Comment by gewarren** [comment]
> I will try to fix the lint errors separately.

These are fixed now, so they'll go away the next time the PR builds.

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (JamesNK) [inline docs/core/diagnostics/metrics-instrumentation.md:144]: I'm not sure we'd want to suggest this as an option at all. A new `Controller` instance is created for every request.

I know Meter and counters are cached, so it doesn't matter too much, but it feels clean to "create" these types once in their own type like is suggested immediately below.

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (JamesNK) [inline docs/core/diagnostics/metrics-instrumentation.md:259]: Change to seconds? Change type to long? Also, add "s" as unit

Comment (JamesNK) [reply docs/core/diagnostics/metrics-instrumentation.md:259]: I see this in recommendations:

> For timing things, Histogram is usually preferred. Often it's useful to understand the tail of these distributions (90th, 95th, 99th percentile) rather than averages or totals.

Recommend seconds?

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (JamesNK) [inline docs/core/diagnostics/metrics-instrumentation.md:598]: Should `MetricCollector` have a ctor that doesn't have a scope argument?

Comment (noahfalk) [reply docs/core/diagnostics/metrics-instrumentation.md:598]: It does have ctors like (Meter, string) and (Instrument), but it was intentional not to offer a ctor that takes a meter name string binds in global scope by default.

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (noahfalk) [reply docs/core/diagnostics/metrics-instrumentation.md]: Removed

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (noahfalk) [reply docs/core/diagnostics/metrics-instrumentation.md]: I changed it to seconds, but didn't include seconds in the unit because we haven't reached a discussion of units in the doc yet.

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (JamesNK) [inline docs/core/diagnostics/metrics-instrumentation.md:380]:
```suggestion
braces around "{hats}" is part of the UCUM standard, indicating that it is a descriptive annotation rather than a unit name with a standardized meaning like seconds or bytes.
```
To avoid confusion about the different meanings of meter.

Comment (noahfalk) [reply docs/core/diagnostics/metrics-instrumentation.md:380]: good call... I didn't even notice that.

**Comment by noahfalk** [comment]
OK I switched this to no longer be a draft. I'll wait until end of Friday to merge in case anyone has more comments on the parts I added today. Thanks all!

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (tommcdon) [inline docs/core/diagnostics/metrics-instrumentation.md:22]: Should we update '.NET Core 3.1' to '.NET 6'?  .NET Core 3.1 is no longer supported.

File: docs/core/diagnostics/metrics-instrumentation.md
Comment (tommcdon) [inline docs/core/diagnostics/metrics-instrumentation.md:124]: Should we mention that Meter DI was added in .NET 8?

Comment (tommcdon) [reply docs/core/diagnostics/metrics-instrumentation.md:124]: Please disregard - I see that .NET 8 was mentioned earlier.

**Pull request merged**
